crates/bevy_ecs/src/world/mod.rs

```diff
@@ -2635,6 +2635,212 @@ impl World {
+                .components()
+                .get_name(component_id)
+                .map(ToString::to_string);
+
+            return Err(GetComponentReflectError::EntityDoesNotHaveComponent {
+                entity,
+                type_id,
+                component_id,
+                component_name,
+            });
+        };
+
+        let Some(type_registry) = self.get_resource::<AppTypeRegistry>().map(|atr| atr.read())
+        else {
+            return Err(GetComponentReflectError::MissingAppTypeRegistry);
+        };
+
+        let Some(reflect_from_ptr) = type_registry.get_type_data::<ReflectFromPtr>(type_id) else {
+            return Err(GetComponentReflectError::MissingReflectFromPtrTypeData(
+                type_id,
+            ));
+        };
+
+        // SAFETY:
+        // - `comp_ptr` is guaranteed to point to an object of type `type_id`
+        // - `reflect_from_ptr` was constructed for type `type_id`
+        // - Assertion that checks this equality is present
+        unsafe {
+            assert_eq!(
+                reflect_from_ptr.type_id(),
+                type_id,
+                "Mismatch between Ptr's type_id and ReflectFromPtr's type_id",
+            );
+
+            Ok(reflect_from_ptr.as_reflect(comp_ptr))
+        }
+    }
+
+    /// Retrieves a mutable reference to the given `entity`'s [`Component`] of the given `type_id` using
+    /// reflection.
+    ///
+    /// Requires [`#[derive(Reflect)`](derive@bevy_reflect::Reflect) on the [`Component`] and
+    /// `app.register_type::<TheComponent>()` to have been called.
+    ///
+    /// This is the mutable version of [`World::get_reflect()`], see its docs for more information
+    /// and an example.
+    ///
+    /// Just calling this method does not trigger [change detection](crate::change_detection).
+    ///
+    /// # Errors
+    ///
+    /// See [`GetComponentReflectError`](error::GetComponentReflectError) for the possible errors and their descriptions.
+    ///
+    /// # Example
+    ///
+    /// See the documentation for [`World::get_reflect()`].
+    ///
+    /// # Note
+    /// Requires the feature `bevy_reflect` (included in the default features).
+    #[inline]
+    #[cfg(feature = "bevy_reflect")]
+    pub fn get_reflect_mut(
+        &mut self,
+        entity: Entity,
+        type_id: TypeId,
+    ) -> Result<Mut<'_, dyn bevy_reflect::Reflect>, error::GetComponentReflectError> {
+        use bevy_reflect::ReflectFromPtr;
+
+        use crate::prelude::AppTypeRegistry;
+
+        use error::GetComponentReflectError;
+
+        // little clone() + read() dance so we a) don't keep a borrow of `self` and b) don't drop a
+        // temporary (from read()) too  early.
+        let Some(app_type_registry) = self.get_resource::<AppTypeRegistry>().cloned() else {
+            return Err(GetComponentReflectError::MissingAppTypeRegistry);
+        };
+        let type_registry = app_type_registry.read();
```

Comment:
It essentially clones an `Arc`, so it's not a huge clone. The clone is necessary because otherwise we keep `self` borrowed immutably, which will prevent the call to `self.get_mut_by_id()` later one, because that needs a mutable borrow (which it can't get then). That's also why it isn't necessary in the immutable version of the function.

But since it's just a clone of an `Arc` essentially, which is kind of meant to be cloned, I didn't worry too much about it. Tried to make it clear in the comment above, but I probably should just have said "we clone here because otherwise we keep an immutable borrow, which will prevent calling `self.get_mut_by_id()`" :sweat_smile: . I'll try to clear that up.

Reply:
yeah the performance is ok, i just forgot `read` held a reference to the `RwLock` at all. perfectly fine to leave as is.